From 297713c8952c9e02e41013c833c8becaa1bce279 Mon Sep 17 00:00:00 2001
From: "Nguyen Van Linh [FGA.BU13]" <LinhNv25@fsoft.com.vn>
Date: Mon, 9 Apr 2018 23:38:34 +0700
Subject: [PATCH 271/641] serial: sh-sci: Add lock to prevent the SCSCR
 register access conflict

	Commit: <6f96fbf5ba9ac18e1e50e7ebd62b7dc29140fd10>
	From:	git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-backport.git

	sci_start_tx() changes the SCSCR register, but does not lock sci_port.
	Therefore, a caller should lock sci_port.

	sci_tx_dma_release() does not lock sci_port in calling sci_start_tx().
	This may cause conflicts of settings to the SCSCR register.
	So, this patch adds lock in sci_tx_dma_release().

Signed-off-by: Mitsuru Ezo <mitsuru.ezo.pz@ps.hitachi-solutions.com>
Signed-off-by: Kazuya Mizuguchi <kazuya.mizuguchi.ks@renesas.com>
Signed-off-by: tungvo <tung.vo.eb@rvc.renesas.com>
Signed-off-by: Nguyen Van Linh [FGA.BU13] <LinhNv25@fsoft.com.vn>
Signed-off-by: vietn <vietn@fsoft.com.vn>
---
 drivers/tty/serial/sh-sci.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/tty/serial/sh-sci.c b/drivers/tty/serial/sh-sci.c
index 20d2e72..65cdb7b 100644
--- a/drivers/tty/serial/sh-sci.c
+++ b/drivers/tty/serial/sh-sci.c
@@ -1192,8 +1192,11 @@ static void sci_tx_dma_release(struct sci_port *s, bool enable_pio)
 		dma_release_channel(chan);
 	}
 
-	if (enable_pio)
+	if (enable_pio) {
+		spin_lock_irqsave(&port->lock, flags);
 		sci_start_tx(port);
+		spin_unlock_irqrestore(&port->lock, flags);
+	}
 }
 
 static void sci_submit_rx(struct sci_port *s)
-- 
2.7.4

